core/src/main/scala/org/http4s/Query.scala

```diff
@@ -1,13 +1,82 @@
 package org.http4s
 
-import org.http4s.Query._
+import org.http4s.FormQuery._
 import org.http4s.parser.QueryParser
-import org.http4s.util.{UrlFormCodec, UrlCodingUtils, Writer, Renderable}
+import org.http4s.util._
+import org.http4s.util.encoding.UriCodingUtils
 
 import scala.collection.generic.CanBuildFrom
-import scala.collection.immutable.{BitSet, IndexedSeq}
+import scala.collection.immutable.IndexedSeq
 import scala.collection.mutable.ListBuffer
-import scala.collection.{ IndexedSeqOptimized, mutable }
+import scala.collection.{IndexedSeqOptimized, mutable}
+
+
+sealed trait Query extends QueryOps with Renderable {
+  def asForm: FormQuery
+  def isEmpty: Boolean
+  def encoded: String
+}
+
+/** Represents the absence of a query component */
+object EmptyQuery extends Query {
+  override def asForm: FormQuery = FormQuery()
+  override def encoded: String = "<please check isEmpty before using this value>"
+  override def isEmpty: Boolean = true
+
+  /////////////////////// QueryOps methods and types /////////////////////////
+  override protected type Self = FormQuery
+  override protected def self: Self = formQuery
+  override protected def replaceQuery(query: FormQuery): Self = query
+  override protected val formQuery: FormQuery = asForm
+  ////////////////////////////////////////////////////////////////////////////
+
+  /** Base method for rendering this object efficiently */
+  override def render(writer: Writer): writer.type = writer << "EmptyQuery"
+}
+
+object Query {
+  def apply(s: String): PlainQuery = new PlainQuery(s)
+  def apply(xs: (String, Option[String])*): FormQuery = FormQuery(xs: _*)
+
+  val empty: Query = EmptyQuery
+
+  def fromString(s: String): PlainQuery = new PlainQuery(s)
+  def fromPairs(xs: (String, String)*): FormQuery = FormQuery.fromPairs(xs: _*)
+  def fromMap(map: Map[String, Seq[String]]): FormQuery = FormQuery.fromMap(map)
+
+  def equals(q1: Query, q2: Query): Boolean =
+    (q1.isEmpty && q2.isEmpty) || (!q1.isEmpty && !q2.isEmpty && q1.encoded == q2.encoded)
+
+}
+
+final case class PlainQuery(plain: String) extends Query with QueryOps with Renderable {
```

Comment:
Do these type constructors need to be public, or is this an implementation detail?